### ISSUE TITLE
provisioners receive incomplete configuration during validate

### ISSUE THREAD

**Issue by radekg** [hashicorp/terraform#17439]
I am attempting to use the Ansible provisioner plugin I've created over the last few days and I believe I found some really weird edge case where the provisioner configuration is not properly interpolated.

The plugin: https://github.com/radekg/terraform-provisioner-ansible

What exactly happens: in my plugin, I validate the existence of the playbook file before the module runs. This check happens here: https://github.com/radekg/terraform-provisioner-ansible/blob/master/provisioner.go#L575

When I call my provisioner in the module like this:

```
    plays {
      playbook = "${path.module}/../../../ansible/playbook-k8s-controller.yml"
      extra_vars {
        etcd_servers  = "${join(",", data.template_file.etcd_server_address.*.rendered)}"
      }
    }
```

I get an error saying that the file (literal):

    ${path.module}/../../../ansible/playbook-k8s-controller.yml` does not exist

If I remove the line with `etcd_servers` `join` interpolation but change the plays path (to trigger the error) to `${path.module}/../../../meh/ansible/playbook-k8s-controller.yml`, the error is

    /Users/rad/dev/kubernetes/terraform/controllers/../../../meh/ansible/playbook-k8s-controller.yml  does not exist

In other words, from what I'm observing here, the former case, with the `${join(...)}` does not seem to be interpolating `${path.module}`. This only happens at the validation stage, apply completes fine. I'd expect that no matter what, my `${path.module}` would get interpolated when validating.

I guess I'm in such an edge case here that it's better to await for more questions rather than doing any guesswork.

Terraform version 0.11.3, plugin is built against 0.11.3.

There is no crash in the plugin.

**Comment by apparentlymart** [comment]
Hi @radekg! Sorry for this annoying behavior.

As currently implemented, it is expected that a provisioner gets the raw (not interpolated) configuration during its validate step. The expectations here are described in [the `Validate` function documentation](https://godoc.org/github.com/hashicorp/terraform/terraform#ResourceProvisioner):

```
    // Validate is called once at the beginning with the raw
    // configuration (no interpolation done) and can return a list of warnings
    // and/or errors.
    //
    // This is called once per resource.
    //
    // This should not assume any of the values in the resource configuration
    // are valid since it is possible they have to be interpolated still.
    // The primary use case of this call is to check that the required keys
    // are set and that the general structure is correct.
    Validate(*ResourceConfig) ([]string, []error)
```

As part of switching to the new configuration interpreter implementation (which is our current focus) we'll be able to start passing more information in here, but it'll still be incomplete just because at validation time we don't yet know all possible values.

The good news is that we probably _will_ be able to resolve `path.module` there because that reference is trivial to resolve at validation time. Your `etcd_servers` argument, on the other hand, will not be known yet because it refers to a data resource and those aren't read until a later step. I'm not sure at this time if this will be possible immediately after our current work is complete or if we will need to make some further adjustments to the provisioner-handling code, but we do at least intend to provide information as complete as possible here.

For now, `validate` is focused only on validating the _keys_ in the configuration, so the change you made to remove that check is the right thing to do for the moment.

**Comment by alicankustemur** [comment]
in currently i have a solution and it works.
`
variable "path" {
    type = "string"
    default = "~/terraform-project/"
}

provisioner "local-exec" {
    ommand = "cp -r ${var.path}/ansible ${path.module}/"
}

provisioner "ansible" {
    plays {
      playbook = "${path.module}/ansible/playbook.yaml"
      hosts = ["all"]
    }
}

provisioner "local-exec" {
    command = "rm -r ${path.module}/ansible"
} 
`
first copy file to under `path.module` and runs playbook then remove this folder.
so i have one playbook folder and i apply my changes on this folder.

**Comment by jbardin** [comment]
As of 0.15 the internal provisioners will be using the full type system available in terraform, and will no longer see the legacy raw interpolation values. While any remaining external provisioner plugins will be getting more accurate information in releases after 0.12, their SDK is frozen with no further changes being planned. 

Any changes to the provisioner plugin SDK will be done along with the providers, and the issue is being tracked in https://github.com/hashicorp/terraform-plugin-sdk/issues/58

Thanks!
